main.go

```diff
@@ -47,38 +67,80 @@ func main() {
+	var metrics = &metrics{}
+	collectors, err := prepareCollectors(opts)
+	if err != nil {
+		log.Fatalf("[ERROR] Can't initialise collectors, %v", err)
+	}
+	senders, err := prepareSenders(opts)
+	if err != nil {
+		log.Fatalf("[ERROR] Can't initialise senders, %v", err)
 	}
 
+	for ticker := time.NewTicker(opts.CollectPeriod); true; <-ticker.C {
+		collectMetrics(metrics, collectors, "https://status.cloud.google.com/incidents.json")
+		sendMetrics(metrics, senders, opts)
+	}
+}
+
+// create and return a link to collectors with credentials provided via opts,
+// return error in case of problems with connection initialisation
+func prepareCollectors(opts opts) (*collectors, error) {
+	var collectors = &collectors{}
+	var err error
 	if opts.PrismAPIKey != "" && opts.PrismAPIPassword != "" {
-		log.Print("[INFO] Starting Prisma data collection")
-		prisma, err := api.NewPrisma(opts.PrismAPIKey, opts.PrismAPIPassword, opts.PrismAPIUrl)
+		log.Print("[INFO] Initialising Prisma data collection")
+		collectors.prisma, err = api.NewPrisma(opts.PrismAPIKey, opts.PrismAPIPassword, opts.PrismAPIUrl)
 		if err != nil {
-			log.Fatalf("[ERROR] Can't connect to Prisma, %v", err)
+			return nil, errors.Wrap(err, "can't connect to Prisma")
 		}
-		collectedMetrics.complianceInfo, err = prisma.GatherComplianceInfo()
+	}
+	if opts.SCCOrgID != "" {
+		log.Print("[INFO] Initialising Google Security Command Center data collection")
+		collectors.sccSources, err = api.GetSCCSourcesByName(opts.SCCOrgID, opts.SCCSourcesRegex)
 		if err != nil {
-			log.Printf("[ERROR] Can't request complience information, %v", err)
+			return nil, errors.Wrap(err, "can't get SCC sources information")
 		}
-		collectedMetrics.prismaHealthStatus = prisma.GetAPIHealthStatus()
 	}
-	collectedMetrics.googleSCCHealthStatus = api.GetSCCHealthStatus("https://status.cloud.google.com/incidents.json")
+	return collectors, nil
+}
 
-	Graphite := graphite.New(opts.GraphiteHost, opts.GraphitePort, opts.GraphitePrefix)
-	if err := graphite.SendComplianceInfo(Graphite, opts.CompliancePrefix, collectedMetrics.complianceInfo); err != nil {
-		log.Printf("[ERROR] Can't send complience information, %v", err)
+// create and return a pointer to senders
+func prepareSenders(opts opts) (*senders, error) {
+	var err error
+	var senders = &senders{}
+	if opts.GraphiteHost != "" {
+		if senders.graphite, err = graphite.New(opts.GraphiteHost, opts.GraphitePort, opts.GraphitePrefix); err != nil {
+			return nil, errors.Wrap(err, "can't create Graphite")
+		}
 	}
-	if opts.SCCOrgID != "" {
-		sccSources, err := api.GetSCCSourcesByName(opts.SCCOrgID, opts.SCCSourcesRegex)
-		if err != nil {
-			log.Fatalf("[ERROR] Can't get SCC sources information, %v", err)
+	return senders, nil
+}
+
+// collectMetrics collects metrics into referenced metrics object using provided collectors
+func collectMetrics(metrics *metrics, collectors *collectors, googleHealthDashboard string) {
+	var err error
+	if collectors.prisma != nil {
```

Comment:
nil pointer error if collectors is nil. Since the [code here](https://github.com/bookingcom/cloudsec-metrics/pull/2/files#diff-7ddfb3e035b42cd70649cc33393fe32cR71-R73) is not stopping execution, this is likely to happen.